include/watchdog.h

```diff
@@ -1,116 +1,168 @@
-};
+/**
+ * @brief Watchdog operational mode.
+ */
+#define WDT_OP_MODE_RESET		(0)
+#define WDT_OP_MODE_INTERRUPT_RESET	BIT(0)
+#define WDT_OP_MODE_MASK		(0x1)
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog behavior when CPU is in sleep state.
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+#define WDT_NO_PAUSE_IN_SLEEP	(0)
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+#define WDT_PAUSE_IN_SLEEP_MASK (1 << 1)
 
+/**
+ * @brief Watchdog behavior when CPU is halted by the debugger.
+ */
+#define WDT_NO_PAUSE_HALTED_BY_DBG	(0)
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+#define WDT_PAUSE_HALTED_BY_DBG_MASK	(1 << 2)
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog configuration struct.
+ *
+ * timeouts - pointer to an array storing timeouts values in microseconds
+ *	      for all enabled channels. If watchdog supports multiple channels
+ *	      but only one counter reload value then the first value from the
+ *	      array will be taken.
+ * interrupt_fn - function that is going to be called if
+ *		  WDT_OP_MODE_INTERRUPT_RESET is enabled.
+ * operation - a bitfield with following parts:
+ *    operational mode    [ 0 ]       - reset or interrupt -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
+ * num_of_channels - number of channels that should be enabled on
+ *		     given watchdog instance.
  */
 struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+	u32_t *timeouts;
+	void (*interrupt_fn)(struct device *dev, u8_t channel);
+	u8_t operation;
+	u8_t num_of_channels;
 };
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
+typedef int (*wdt_api_enable)(struct device *dev);
+typedef int (*wdt_api_disable)(struct device *dev);
 typedef int (*wdt_api_set_config)(struct device *dev,
 				  struct wdt_config *config);
 typedef void (*wdt_api_get_config)(struct device *dev,
 				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
+typedef void (*wdt_api_reload)(struct device *dev, u8_t channel);
 
 struct wdt_driver_api {
 	wdt_api_enable enable;
 	wdt_api_disable disable;
-	wdt_api_get_config get_config;
 	wdt_api_set_config set_config;
+	wdt_api_get_config get_config;
 	wdt_api_reload reload;
 };
 
-static inline void wdt_enable(struct device *dev)
+/*
+ * @brief Enable watchdog instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance has not been configured yet.
+ * @retval -EBUSY if watchdog instance has been already enabled.
+ */
+static inline int wdt_enable(struct device *dev)
 {
```

Comment:
isn't wdt_enable/wdt_disable something related to PM? (actual implementation in wdt_qmsi seems to ack that - qmsi, against which the legacy wdt api was designed from.)

Looks like you can remove these 2 functions, up to PM to handle that.

Reply:
Not all wdt controller seem to be possible to disable once started, at least not through the controller itself. And disabling through wdt_configure would look weird.

Maybe we need these 2 after all, and the PM function would redirect to them if enabled.

Reply:
Considering the fact that we don't need to pass any parameters when we want to disable the watchdog it may be indeed a cleaner / more user friendly design if we have a dedicated wdt_disable function rather than passing some flag to wdt_configure. But we likely won't need wdt_enable.